mmdet/models/utils/transformer.py

```diff
@@ -0,0 +1,725 @@
+                memory_attn_mask=None,
+                target_attn_mask=None,
+                memory_key_padding_mask=None,
+                target_key_padding_mask=None):
+        """Forward function for `TransformerDecoder`.
+
+        Args:
+            x (Tensor): Input query. Same in `TransformerDecoderLayer.forward`.
+            memory (Tensor): Same in `TransformerDecoderLayer.forward`.
+            memory_pos (Tensor): Same in `TransformerDecoderLayer.forward`.
+                Default None.
+            query_pos (Tensor): Same in `TransformerDecoderLayer.forward`.
+                Default None.
+            memory_attn_mask (Tensor): Same in
+                `TransformerDecoderLayer.forward`. Default None.
+            target_attn_mask (Tensor): Same in
+                `TransformerDecoderLayer.forward`. Default None.
+            memory_key_padding_mask (Tensor): Same in
+                `TransformerDecoderLayer.forward`. Default None.
+            target_key_padding_mask (Tensor): Same in
+                `TransformerDecoderLayer.forward`. Default None.
+
+        Returns:
+            Tensor: forwarded results from `TransformerDecoder`.
+        """
+        intermediate = []
+        for layer in self.layers:
+            x = layer(x, memory, memory_pos, query_pos, memory_attn_mask,
+                      target_attn_mask, memory_key_padding_mask,
+                      target_key_padding_mask)
+            if self.return_intermediate:
+                intermediate.append(self.norm(x))
+        if self.norm is not None:
+            x = self.norm(x)
+            if self.return_intermediate:
+                intermediate.pop()
+                intermediate.append(x)
+        if self.return_intermediate:
+            return torch.stack(intermediate)
+        return x.unsqueeze(0)
+
+    def __repr__(self):
+        """str: a string that describes the module"""
+        repr_str = self.__class__.__name__
+        repr_str += f'(num_layers={self.num_layers}, '
+        repr_str += f'embed_dims={self.embed_dims}, '
+        repr_str += f'num_heads={self.num_heads}, '
+        repr_str += f'feedforward_channels={self.feedforward_channels}, '
+        repr_str += f'dropout={self.dropout}, '
+        repr_str += f'order={self.order}, '
+        repr_str += f'act_cfg={self.act_cfg}, '
+        repr_str += f'norm_cfg={self.norm_cfg}, '
+        repr_str += f'num_fcs={self.num_fcs}, '
+        repr_str += f'return_intermediate={self.return_intermediate})'
+        return repr_str
+
+
+class Transformer(nn.Module):
+    """Implements the DETR transformer.
+
+    Following the official DETR implementation, this module copy-paste
+    from torch.nn.Transformer with modifications:
+        * positional encodings are passed in MultiheadAttention
+        * extra LN at the end of encoder is removed
+        * decoder returns a stack of activations from all decoding layers
+
+    See 'End-to-End Object Detection with Transformers'
```

Comment:
`'End-to-End Object Detection with Transformers'
    (https://arxiv.org/pdf/2005.12872)` -> 
```
`paper: End-to-End Object Detection with Transformers
    <https://arxiv.org/pdf/2005.12872>`_
```